File: integrations/openclaw/src/dkgClient.ts
/**
 * Minimal fetch-based client for the local DKG v10 node HTTP API.
 *
 * Zero deps — uses global `fetch` (Node ≥18) and `node:fs` for token loading.
 * Every method fails open: on any transport/HTTP error it throws `DkgError`,
 * and callers treat that as "node unreachable, carry on" (never break the
 * agent loop).
 *
 * URL/token resolution mirrors the hermes `dkg_client.py`:
 *   url   ← opts.url | $DKG_DAEMON_URL | http://127.0.0.1:9200
 *   token ← opts.token | $DKG_API_TOKEN | $DKG_AUTH_TOKEN | $DKG_HOME/auth.token | ~/.dkg/auth.token
 */
import { readFileSync } from "node:fs";
import { homedir } from "node:os";
import { join } from "node:path";
import type { Quad } from "./quads.js";

export class DkgError extends Error {
  readonly status?: number;
  constructor(message: string, status?: number) {
    super(message);
    this.name = "DkgError";
    this.status = status;
  }
}

const DEFAULT_URL = "http://127.0.0.1:9200";
const DEFAULT_TIMEOUT_MS = 3000;
// SPARQL reads fan out across every shared-memory asset, so a large curated
// graph can take a few seconds. Only the background ruleset sync hits this, so a
// generous ceiling is safe.
const QUERY_TIMEOUT_MS = 30000;

export function resolveDkgUrl(explicit?: string): string {
  return (explicit || process.env.DKG_DAEMON_URL || DEFAULT_URL).replace(/\/+$/, "");
}

export function resolveDkgToken(explicit?: string): string | undefined {
  if (explicit) return explicit;
  const env = process.env.DKG_API_TOKEN || process.env.DKG_AUTH_TOKEN;
  if (env) return env;
  const home = process.env.DKG_HOME || join(homedir(), ".dkg");
  try {
    // The token file has a leading comment line (e.g. "# DKG node API token —
    // ..."); return the first non-comment, non-blank line, matching the Python
    // client. Using the whole file would put non-ASCII comment text into the
    // Authorization header and throw a ByteString error.
    for (const raw of readFileSync(join(home, "auth.token"), "utf8").split(/\r?\n/)) {
      const line = raw.trim();
      if (line && !line.startsWith("#")) return line;
    }
  } catch {
    // no token file — node may not require one on local hardhat.
  }
  return undefined;
}

export interface DkgClientOptions {
  url?: string;
  token?: string;
  timeoutMs?: number;
}

export type DkgView = "working-memory" | "shared-working-memory" | "verifiable-memory";

export class DkgClient {
  readonly url: string;
  private readonly token?: string;
  private readonly timeoutMs: number;
  /** Cached reporter agent address (see `reporterAddress()`). */
  private cachedReporter?: string;

  constructor(opts: DkgClientOptions = {}) {
    this.url = resolveDkgUrl(opts.url);
    this.token = resolveDkgToken(opts.token);
    this.timeoutMs = opts.timeoutMs ?? DEFAULT_TIMEOUT_MS;
  }

  private async request<T = unknown>(
    path: string,
    body?: unknown,
    method = "POST",
    timeoutMs = this.timeoutMs,
  ): Promise<T> {
    const controller = new AbortController();
    const timer = setTimeout(() => controller.abort(), timeoutMs);
    try {
      const headers: Record<string, string> = { "content-type": "application/json" };
      if (this.token) headers.authorization = `Bearer ${this.token}`;
      const res = await fetch(`${this.url}${path}`, {
        method,
        headers,
        body: body === undefined ? undefined : JSON.stringify(body),
        signal: controller.signal,
      });
      const text = await res.text();
      if (!res.ok) {
        throw new DkgError(`DKG ${method} ${path} -> ${res.status}: ${text.slice(0, 300)}`, res.status);
      }
      return (text ? JSON.parse(text) : {}) as T;
    } catch (err) {
      if (err instanceof DkgError) throw err;
      throw new DkgError(`DKG ${method} ${path} failed: ${(err as Error).message}`);
    } finally {
      clearTimeout(timer);
    }
  }

  /** Node liveness probe — public, no-auth `GET /api/status`. */
  async status(): Promise<unknown> {
    return this.request("/api/status", undefined, "GET");
  }

  /**
   * Resolve the node's reporter agent address, cached. Prefers the definitive
   * `GET /api/agent/identity` → `{agentAddress}`; falls back to `GET /api/status`
   * (which may surface an agent address). Fails open to "node" so a report URI
   * can always be namespaced. NEVER derived from the auth token string.
   */
  async reporterAddress(): Promise<string> {
    if (this.cachedReporter !== undefined) return this.cachedReporter;
    const pick = (resp: unknown): string | undefined => {
      const r = resp as Record<string, unknown> | null;
      const addr = r?.agentAddress ?? r?.agent_address ?? r?.address;
      return typeof addr === "string" && addr.trim() ? addr.trim() : undefined;
    };
    try {
      const addr = pick(await this.request("/api/agent/identity", undefined, "GET"));
      if (addr) return (this.cachedReporter = addr);
    } catch {
      // definitive route unavailable — fall through to /api/status.
    }
    try {
      const addr = pick(await this.status());
      if (addr) return (this.cachedReporter = addr);
    } catch {
      // node unreachable — fail open below.
    }
    return (this.cachedReporter = "node");
  }

  /**
   * One-shot create + write + seal (+ share to SWM). This is the report path:
   * `POST /api/knowledge-assets {contextGraphId, name, quads, alsoShareSwm:true}`.
   * Quads carry NO per-quad graph (the CG scopes them).
   */
  async shareKnowledgeAsset(
    contextGraphId: string,
    name: string,
    quads: Quad[],
  ): Promise<{ kaId?: string; ual?: string } & Record<string, unknown>> {
    return this.request("/api/knowledge-assets", {
      contextGraphId,
      name,
      quads,
      alsoShareSwm: true,
    });
  }

  /** SPARQL query against a memory view (ruleset sync reads shared-working-memory). */
  async query(
    sparql: string,
    contextGraphId: string,
    view: DkgView = "shared-working-memory",
  ): Promise<{ results?: unknown } & Record<string, unknown>> {
    return this.request("/api/query", { sparql, contextGraphId, view }, "POST", QUERY_TIMEOUT_MS);
  }
}

File: plugins/guardian/dkg_client.py
"""Stdlib HTTP client for the local DKG v10 node.

Wraps the daemon's write/query API with a tiny, dependency-free ``urllib``
client. URL and token resolution mirror the original telemetry plugin
(``DKG_DAEMON_URL`` env → default; ``DKG_API_TOKEN``/``DKG_AUTH_TOKEN`` env →
``$DKG_HOME/auth.token``). Every request uses a short timeout and raises
:class:`DkgError` on a non-2xx response; all callers fail open.
"""

from __future__ import annotations

import json
import logging
import os
import urllib.error
import urllib.request
from pathlib import Path
from typing import Any, Dict, List, Optional

from . import constants

logger = logging.getLogger(__name__)

_TIMEOUT = 3.0
#: SPARQL reads fan out across every shared-memory asset in the graph, so a
#: large curated graph can take a few seconds to evaluate. This is only ever
#: hit by the background ruleset refresh and the dashboard — never the hot hook
#: path, which serves a cached ruleset — so a generous ceiling is safe.
_QUERY_TIMEOUT = 30.0
#: On-chain ops (CG register, VM publish) need block confirmation — far longer
#: than the short read timeout. Used only by :meth:`register_context_graph` and
#: :meth:`publish`.
_ONCHAIN_TIMEOUT = 180.0
Quad = Dict[str, str]


class DkgError(RuntimeError):
    """Raised for any non-2xx daemon response or transport failure."""


def _is_already_finalized(exc: DkgError) -> bool:
    """True when a share failed only because the KA already exists sealed.

    The daemon rejects re-finalizing an assertion with a different (or same)
    merkle root; for our deterministic threat/report names that just means the
    content is already on the graph, so the caller can treat it as success.
    """
    msg = str(exc).lower()
    return "already finalized" in msg or "already exists" in msg


def _dkg_home() -> Path:
    env = os.environ.get("DKG_HOME")
    return Path(env).expanduser() if env else Path.home() / ".dkg"


def load_daemon_url() -> str:
    """Resolve the daemon URL: env override → hermes ``dkg.json`` → default."""
    env = os.environ.get("DKG_DAEMON_URL") or os.environ.get("GUARDIAN_DKG_DAEMON_URL")
    if env and env.strip():
        return env.strip().rstrip("/")
    cfg_path = constants.hermes_home() / "dkg.json"
    if cfg_path.exists():
        try:
            data = json.loads(cfg_path.read_text(encoding="utf-8"))
            value = data.get("daemon_url") or data.get("daemonUrl")
            if isinstance(value, str) and value.strip():
                return value.strip().rstrip("/")
        except Exception:
            pass
    return constants.DEFAULT_DKG_URL


def load_token() -> Optional[str]:
    """Resolve the bearer token: env override → ``$DKG_HOME/auth.token``."""
    env = os.environ.get("DKG_API_TOKEN") or os.environ.get("DKG_AUTH_TOKEN")
    if env and env.strip():
        return env.strip()
    token_path = _dkg_home() / "auth.token"
    try:
        if token_path.exists():
            for line in token_path.read_text(encoding="utf-8").splitlines():
                line = line.strip()
                if line and not line.startswith("#"):
                    return line
    except Exception:
        return None
    return None


class DkgClient:
    """Minimal DKG v10 HTTP client. Construct with an explicit url/token or
    let :meth:`from_env` resolve them."""

    def __init__(self, url: Optional[str] = None, token: Optional[str] = None) -> None:
        self.url = (url or load_daemon_url()).rstrip("/")
        self.token = token if token is not None else load_token()

    @classmethod
    def from_env(cls) -> "DkgClient":
        return cls()

    # -- transport ---------------------------------------------------------

    def _request(
        self,
        method: str,
        path: str,
        body: Optional[Dict[str, Any]] = None,
        timeout: Optional[float] = None,
    ) -> Dict[str, Any]:
        data = json.dumps(body).encode("utf-8") if body is not None else None
        headers = {"Accept": "application/json"}
        if data is not None:
            headers["Content-Type"] = "application/json"
        if self.token:
            headers["Authorization"] = f"Bearer {self.token}"
        req = urllib.request.Request(f"{self.url}{path}", data=data, headers=headers, method=method)
        try:
            with urllib.request.urlopen(req, timeout=timeout or _TIMEOUT) as resp:
                raw = resp.read()
        except urllib.error.HTTPError as exc:
            detail = exc.read(1024).decode("utf-8", errors="replace")
            raise DkgError(f"{method} {path} -> {exc.code}: {detail}") from exc
        except (urllib.error.URLError, OSError, TimeoutError) as exc:
            raise DkgError(f"{method} {path} transport error: {exc}") from exc
        if not raw:
            return {}
        try:
            return json.loads(raw.decode("utf-8"))
        except Exception:
            return {}

    # -- status ------------------------------------------------------------

    def status(self) -> Dict[str, Any]:
        """Return node status; raises :class:`DkgError` if unreachable.

        ``GET /api/status`` is the public (no-auth) liveness endpoint; older
        daemons are probed via ``/api/info`` then ``/api/health``.
        """
        for route in ("/api/status", "/api/info", "/api/health"):
            try:
                return self._request("GET", route)
            except DkgError:
                continue
        raise DkgError("node unreachable on /api/status, /api/info, /api/health")

    def agent_identity(self) -> Dict[str, Any]:
        """Resolve the calling token to its agent identity.

        ``GET /api/agent/identity`` → ``{agentAddress, agentDid, name, ...}`` —
        the definitive way to learn which agent the node sees us as.
        """
        return self._request("GET", "/api/agent/identity")

    def reachable(self) -> bool:
        try:
            self.status()
            return True
        except DkgError:
            return False

    # -- context graph -----------------------------------------------------

    def create_context_graph(self, cg_id: str, name: str, description: str = "") -> Dict[str, Any]:
        """Create a local context graph (free, off-chain)."""
        body: Dict[str, Any] = {"id": cg_id, "name": name}
        if description:
            body["description"] = description
        return self._request("POST", "/api/context-graph/create", body)

    def register_context_graph(self, cg_id: str, access_policy: int, publish_policy: int) -> Dict[str, Any]:
        """Register a CG on-chain with explicit access/publish policies.

        For the public curated KB: ``access_policy=0`` (open discovery/share)
        and ``publish_policy=0`` (only curator wallets publish to VM). We pass
        ``publish_policy`` explicitly — an open CG otherwise defaults permissive.
        """
        return self._request(
            "POST",
            "/api/context-graph/register",
            {"id": cg_id, "accessPolicy": access_policy, "publishPolicy": publish_policy},
            timeout=_ONCHAIN_TIMEOUT,
        )

    # -- knowledge assets --------------------------------------------------

    def share_knowledge_asset(self, cg_id: str, name: str, quads: List[Quad]) -> Dict[str, Any]:
        """One-shot create+write+seal+share of a KA to SWM.

        ``POST /api/knowledge-assets {contextGraphId, name, quads, alsoShareSwm:true}``.
        Used for outbound sightings and curated-threat authoring.

        Idempotent: threat/report KA names are deterministic (``sha256`` of the
        threat identifier), so re-sharing the same threat is expected. The node
        rejects re-finalizing an existing sealed assertion; we treat that as
        success — the content is already on the graph.
        """
        try:
            return self._request(
                "POST",
                "/api/knowledge-assets",
                {"contextGraphId": cg_id, "name": name, "quads": quads, "alsoShareSwm": True},
            )
        except DkgError as exc:
            if _is_already_finalized(exc):
                return {"name": name, "idempotent": True}
            raise

    def write_private_knowledge_asset(self, cg_id: str, name: str, quads: List[Quad]) -> Dict[str, Any]:
        """Create+write+seal a KA in WM WITHOUT sharing to SWM (private audit)."""
        return self._request(
            "POST",
            "/api/knowledge-assets",
            {"contextGraphId": cg_id, "name": name, "quads": quads, "alsoShareSwm": False},
        )

    def publish(self, cg_id: str, name: str, epochs: int = 1) -> Dict[str, Any]:
        """Mint a sealed KA on-chain (VM). Returns ``{kaId, ual, txHash, ...}``."""
        return self._request(
            "POST",
            f"/api/knowledge-assets/{name}/vm/publish",
            {"contextGraphId": cg_id, "options": {"publishEpochs": max(1, int(epochs))}},
            timeout=_ONCHAIN_TIMEOUT,
        )

    # -- query -------------------------------------------------------------

    def query(
        self,
        sparql: str,
        cg_id: str,
        view: str = constants.VIEW_SHARED_WORKING_MEMORY,
        on_error: Any = None,
    ) -> List[Dict[str, Any]]:
        """Run a SPARQL SELECT and return normalized bindings.

        Each binding is a ``{var: value}`` dict of plain strings (literals and
        IRIs already unwrapped — see :func:`extract_binding`). On transport
        error it returns *on_error* (default ``[]``) rather than raising, since
        read paths fail open. A caller that must tell a genuine *empty* result
        (``[]``) apart from a *failure* passes a sentinel (e.g. ``on_error=None``
        won't collide with a successful empty ``[]``).
        """
        try:
            result = self._request(
                "POST",
                "/api/query",
                {"sparql": sparql, "contextGraphId": cg_id, "view": view},
                timeout=_QUERY_TIMEOUT,
            )
        except DkgError as exc:
            logger.debug("guardian: query failed: %s", exc)
            return [] if on_error is None else on_error
        return normalize_bindings(result)

    def register_agent(self, name: str, framework: str = "hermes") -> Dict[str, Any]:
        """Register a new agent on the node → ``{agentAddress, authToken, ...}``."""
        return self._request("POST", "/api/agent/register", {"name": name, "framework": framework})


# ---------------------------------------------------------------------------
# SPARQL binding normalization
# ---------------------------------------------------------------------------


def extract_binding(value: Any) -> str:
    """Unwrap a single SPARQL binding cell to a plain string.

    Handles the SPARQL-JSON ``{"value": "..."}`` object shape as well as the
    daemon's bare-string shape (IRIs bare, literals ``"..."``, typed literals
    ``"x"^^<...>``, lang literals ``"x"@en``).
    """
    if value is None:
        return ""
    if isinstance(value, dict):
        inner = value.get("value")
        return str(inner) if inner is not None else ""
    if isinstance(value, str):
        if value.startswith('"'):
            i = 1
            while i < len(value):
                if value[i] == '"' and value[i - 1] != "\\":
                    break
                i += 1
            return value[1:i] if i < len(value) else value
        return value
    return str(value)


def normalize_bindings(result: Any) -> List[Dict[str, Any]]:
    """Extract a list of binding rows from any of the daemon's response shapes."""
    if not isinstance(result, dict):
        return []
    rows = None
    if isinstance(result.get("bindings"), list):
        rows = result["bindings"]
    elif isinstance(result.get("results"), dict) and isinstance(result["results"].get("bindings"), list):
        rows = result["results"]["bindings"]
    elif isinstance(result.get("result"), dict) and isinstance(result["result"].get("bindings"), list):
        rows = result["result"]["bindings"]
    if not isinstance(rows, list):
        return []
    return [row for row in rows if isinstance(row, dict)]
